pkg/plugincmdtree/plugins_cache.go

```diff
@@ -0,0 +1,247 @@
+// So, it would use the help command for each sub-command to extract the aliases supported and finally construct
+// the plugin command tree and adds it to cache so that CLI can extract the command chain by parsing the user input
+// against the plugin command tree.
+func (c *cacheImpl) ConstructAndAddTree(plugin *cli.PluginInfo) error {
+	_, exists := c.pluginCommands.CommandTree[plugin.InstallationPath]
+	if exists {
+		return nil
+	}
+
+	pluginCmdTree, err := c.constructPluginCommandTree(plugin)
+	if err != nil {
+		return errors.Wrapf(err, "failed to generate command tree for plugin %q", plugin.Name)
+	}
+	c.pluginCommands.CommandTree[plugin.InstallationPath] = pluginCmdTree
+
+	return c.savePluginCommandTree()
+}
+func (c *cacheImpl) DeleteTree(plugin *cli.PluginInfo) error {
+	_, exists := c.pluginCommands.CommandTree[plugin.InstallationPath]
+	if !exists {
+		return nil
+	}
+
+	delete(c.pluginCommands.CommandTree, plugin.InstallationPath)
+
+	return c.savePluginCommandTree()
+}
+
+func (c *cacheImpl) savePluginCommandTree() error {
+	data, err := yaml.Marshal(c.pluginCommands)
+	if err != nil {
+		return errors.Wrap(err, "failed to marshal plugin command tree")
+	}
+	err = os.WriteFile(GetPluginsCommandTreeCachePath(), data, 0644)
+	if err != nil {
+		return errors.Wrapf(err, "failed to write the command tree to %q file ", GetPluginsCommandTreeCachePath())
+	}
+	return nil
+}
+
+func (c *cacheImpl) constructPluginCommandTree(plugin *cli.PluginInfo) (*CommandNode, error) {
+	if err := c.pluginDocsGenerator(plugin); err != nil {
+		return nil, errors.Wrapf(err, "failed to generate docs for the plugin %q", plugin.Name)
+	}
+	// construct the command tree
+	cmdTreeRoot := NewCommandNode()
+
+	docsDir := getPluginsDocsCachePath()
+	files, err := os.ReadDir(docsDir)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error while reading local plugin command tree directory")
+	}
+	var aliasErrGroup errgroup.Group
+	for _, file := range files {
+		if file.IsDir() {
+			continue
+		}
+
+		// ignore non mark down files
+		if filepath.Ext(file.Name()) != ".md" {
+			continue
+		}
+
+		filename := strings.TrimSuffix(file.Name(), ".md")
+		cmdNames := strings.Split(filename, "_")
+
+		var aliasArgs []string
+		current := cmdTreeRoot
+		for _, cmdName := range cmdNames {
+			if current.Subcommands[cmdName] == nil {
+				current.Subcommands[cmdName] = NewCommandNode()
+			}
+
+			current = current.Subcommands[cmdName]
+			if cmdName != "tanzu" && cmdName != plugin.Name {
+				aliasArgs = append(aliasArgs, cmdName)
+				aliasArgsCopy := make([]string, len(aliasArgs))
+				copy(aliasArgsCopy, aliasArgs)
+				if !current.AliasProcessed {
+					// kickoff the goroutine to add the alias to the command
+					aliasErrGroup.Go(func() error {
+						cmdAlias, aliasErr := getPluginCommandAlias(plugin, aliasArgsCopy)
+						if aliasErr != nil {
+							return aliasErr
+						}
+						current.Aliases = cmdAlias
+						return nil
+					})
```

Comment:
I think we will need to pass the `aliasArgsCopy` and `current` objects to the goroutine otherwise the reference for both might be different within the goroutine at the time of execution.
Reference: https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables

```suggestion
					aliasErrGroup.Go(func(aliasArgsCopy []string, current *CommandNode) error {
						cmdAlias, aliasErr := getPluginCommandAlias(plugin, aliasArgsCopy)
						if aliasErr != nil {
							return aliasErr
						}
						current.Aliases = cmdAlias
						return nil
					}(aliasArgsCopy, current))
```

Reply:
Good point! Thanks @anujc25. As mentioned in the reference we should be good if we declare a new variable with in the loop, we are be good with `aliasArgsCopy` (as we are declaring a new variable for each iteration). So I will update the PR by declaring another variable for `current` and use it in the go routine function because errorgroup `Go` funciton only support `Go( f func() error)` (without parameters). I did test it and its working fine.